crates/sui-bridge/src/utils.rs

```diff
@@ -125,20 +128,65 @@ pub async fn get_eth_signer_client(url: &str, private_key_hex: &str) -> anyhow::
     Ok(SignerMiddleware::new(provider, wallet))
 }
 
-pub async fn publish_coins_return_add_coins_on_sui_action(
-    wallet_context: &mut WalletContext,
+pub async fn publish_and_register_coins_return_add_coins_on_sui_action(
+    wallet_context: &WalletContext,
     bridge_arg: ObjectArg,
-    publish_coin_responses: Vec<SuiTransactionBlockResponse>,
+    token_packages_dir: Vec<PathBuf>,
     token_ids: Vec<u8>,
     token_prices: Vec<u64>,
     nonce: u64,
 ) -> BridgeAction {
-    assert!(token_ids.len() == publish_coin_responses.len());
-    assert!(token_prices.len() == publish_coin_responses.len());
-    let sender = wallet_context.active_address().unwrap();
-    let rgp = wallet_context.get_reference_gas_price().await.unwrap();
+    assert!(token_ids.len() == token_packages_dir.len());
+    assert!(token_prices.len() == token_packages_dir.len());
+    let sui_client = wallet_context.get_client().await.unwrap();
+    let quorum_driver_api = Arc::new(sui_client.quorum_driver_api().clone());
+    let rgp = sui_client
+        .governance_api()
+        .get_reference_gas_price()
+        .await
+        .unwrap();
+
+    let senders = wallet_context.get_addresses();
+    // We want each sender to deal with one coin
+    assert!(senders.len() >= token_packages_dir.len());
+
+    // publish coin packages
+    let mut publish_tokens_tasks = vec![];
+
+    for (token_package_dir, sender) in token_packages_dir.iter().zip(senders.clone()) {
```

Comment:
that zip is a bit strange for me, but all good.

Reply:
landing now, if anything we will fix later

Reply:
nothing to fix, I just though it was an unusual pattern and that is it. It was not worth to mention I guess.
I can elaborate but we can even talk about it just for fun